File: vec.go
// Copyright 2026 The Cockroach Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0

package goodhistogram

import (
	"fmt"
	"strings"
	"sync"

	"github.com/prometheus/client_golang/prometheus"
	prometheusgo "github.com/prometheus/client_model/go"
)

// HistogramVec is a collection of Histograms partitioned by label values.
// It implements prometheus.Collector so the entire vec can be registered
// with a Prometheus registry. Recording is done on the individual
// *Histogram returned by WithLabelValues.
type HistogramVec struct {
	params     Params
	desc       *prometheus.Desc
	labelNames []string

	mu         sync.RWMutex
	histograms map[string]*labeledHistogram
}

type labeledHistogram struct {
	h          *Histogram
	labelPairs []*prometheusgo.LabelPair
}

// NewHistogramVec creates a new HistogramVec. All child histograms share the
// same Params. The desc is created internally from name, help, and labelNames.
func NewHistogramVec(p Params, name, help string, labelNames []string) *HistogramVec {
	return &HistogramVec{
		params:     p,
		desc:       prometheus.NewDesc(name, help, labelNames, nil),
		labelNames: labelNames,
		histograms: make(map[string]*labeledHistogram),
	}
}

// WithLabelValues returns the Histogram for the given label values,
// creating it if it doesn't exist. Panics if the number of values
// doesn't match the number of label names.
func (v *HistogramVec) WithLabelValues(lvs ...string) *Histogram {
	if len(lvs) != len(v.labelNames) {
		panic(fmt.Sprintf(
			"goodhistogram: expected %d label values, got %d",
			len(v.labelNames), len(lvs),
		))
	}
	key := strings.Join(lvs, "\x00")

	v.mu.RLock()
	if lh, ok := v.histograms[key]; ok {
		v.mu.RUnlock()
		return lh.h
	}
	v.mu.RUnlock()

	v.mu.Lock()
	defer v.mu.Unlock()
	if lh, ok := v.histograms[key]; ok {
		return lh.h
	}
	h := New(v.params)
	v.histograms[key] = &labeledHistogram{
		h:          h,
		labelPairs: makeLabelPairs(v.labelNames, lvs),
	}
	return h
}

// DeleteLabelValues removes the Histogram for the given label values.
// Returns true if the entry existed.
func (v *HistogramVec) DeleteLabelValues(lvs ...string) bool {
	key := strings.Join(lvs, "\x00")
	v.mu.Lock()
	defer v.mu.Unlock()
	_, ok := v.histograms[key]
	delete(v.histograms, key)
	return ok
}

// Reset removes all child histograms.
func (v *HistogramVec) Reset() {
	v.mu.Lock()
	defer v.mu.Unlock()
	v.histograms = make(map[string]*labeledHistogram)
}

// Describe implements prometheus.Collector.
func (v *HistogramVec) Describe(ch chan<- *prometheus.Desc) {
	ch <- v.desc
}

// Collect implements prometheus.Collector.
func (v *HistogramVec) Collect(ch chan<- prometheus.Metric) {
	v.mu.RLock()
	defer v.mu.RUnlock()
	for _, lh := range v.histograms {
		ch <- &histogramMetric{
			desc:       v.desc,
			h:          lh.h,
			labelPairs: lh.labelPairs,
		}
	}
}

// histogramMetric implements prometheus.Metric for a single labeled histogram.
type histogramMetric struct {
	desc       *prometheus.Desc
	h          *Histogram
	labelPairs []*prometheusgo.LabelPair
}

func (m *histogramMetric) Desc() *prometheus.Desc { return m.desc }

func (m *histogramMetric) Write(out *prometheusgo.Metric) error {
	snap := m.h.Snapshot()
	out.Histogram = snap.ToPrometheusHistogram()
	out.Label = m.labelPairs
	return nil
}

func makeLabelPairs(names, values []string) []*prometheusgo.LabelPair {
	pairs := make([]*prometheusgo.LabelPair, len(names))
	for i := range names {
		n := names[i]
		v := values[i]
		pairs[i] = &prometheusgo.LabelPair{Name: &n, Value: &v}
	}
	return pairs
}

File: rust/src/vec.rs
// Copyright 2026 The Cockroach Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0

//! A collection of histograms partitioned by label values, mirroring Go's
//! `HistogramVec`. This is the core data structure (no Prometheus dependency);
//! the Prometheus-registerable equivalent is `prometheus::PromHistogramVec`
//! behind the `prometheus` feature.

use crate::{Histogram, Params};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// A set of [`Histogram`]s that share the same [`Params`], keyed by label
/// values. Recording is done on the `Arc<Histogram>` returned by
/// [`HistogramVec::with_label_values`].
pub struct HistogramVec {
    params: Params,
    label_names: Vec<String>,
    inner: RwLock<HashMap<Vec<String>, Arc<Histogram>>>,
}

impl HistogramVec {
    /// Creates a new vec whose children all use `params`.
    pub fn new(params: Params, label_names: &[&str]) -> Self {
        HistogramVec {
            params,
            label_names: label_names.iter().map(|s| s.to_string()).collect(),
            inner: RwLock::new(HashMap::new()),
        }
    }

    /// The label names this vec is partitioned by.
    pub fn label_names(&self) -> &[String] {
        &self.label_names
    }

    /// Returns the histogram for the given label values, creating it if it does
    /// not yet exist. Panics if the number of values doesn't match the number
    /// of label names (matching Go's `WithLabelValues`).
    pub fn with_label_values(&self, lvs: &[&str]) -> Arc<Histogram> {
        self.check_len(lvs);
        let key: Vec<String> = lvs.iter().map(|s| s.to_string()).collect();

        if let Some(h) = self.inner.read().unwrap().get(&key) {
            return Arc::clone(h);
        }
        let mut m = self.inner.write().unwrap();
        // Re-check after taking the write lock.
        if let Some(h) = m.get(&key) {
            return Arc::clone(h);
        }
        let h = Arc::new(Histogram::new(self.params));
        m.insert(key, Arc::clone(&h));
        h
    }

    /// Removes the histogram for the given label values. Returns true if it
    /// existed. Panics on label-count mismatch (matching Go).
    pub fn delete_label_values(&self, lvs: &[&str]) -> bool {
        self.check_len(lvs);
        let key: Vec<String> = lvs.iter().map(|s| s.to_string()).collect();
        self.inner.write().unwrap().remove(&key).is_some()
    }

    /// Removes all child histograms.
    pub fn reset(&self) {
        self.inner.write().unwrap().clear();
    }

    fn check_len(&self, lvs: &[&str]) {
        assert_eq!(
            lvs.len(),
            self.label_names.len(),
            "goodhistogram: expected {} label values, got {}",
            self.label_names.len(),
            lvs.len()
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn params() -> Params {
        Params {
            lo: 1.0,
            hi: 1e6,
            error_bound: 0.1,
        }
    }

    #[test]
    fn same_labels_return_same_histogram() {
        let v = HistogramVec::new(params(), &["op"]);
        let a = v.with_label_values(&["read"]);
        a.record(100);
        let b = v.with_label_values(&["read"]); // same handle
        b.record(200);
        assert_eq!(a.snapshot().total_count, 2);
        assert_eq!(b.snapshot().total_count, 2);

        // A different label set is independent.
        let c = v.with_label_values(&["write"]);
        c.record(300);
        assert_eq!(c.snapshot().total_count, 1);
        assert_eq!(a.snapshot().total_count, 2);
    }

    #[test]
    fn delete_and_reset() {
        let v = HistogramVec::new(params(), &["op"]);
        v.with_label_values(&["read"]).record(1);
        v.with_label_values(&["write"]).record(1);

        assert!(v.delete_label_values(&["read"]));
        assert!(!v.delete_label_values(&["read"])); // already gone

        // A recreated label set starts fresh.
        assert_eq!(v.with_label_values(&["read"]).snapshot().total_count, 0);

        v.reset();
        assert_eq!(v.with_label_values(&["write"]).snapshot().total_count, 0);
    }

    #[test]
    #[should_panic]
    fn wrong_label_count_panics() {
        let v = HistogramVec::new(params(), &["op"]);
        v.with_label_values(&["read", "extra"]);
    }
}
